pkgs/build-support/docker/default.nix

```diff
@@ -994,4 +996,177 @@ rec {
+          [ -e $stdenv/setup ] && source $stdenv/setup
+          PATH=${lib.escapeShellArg (dirOf shell)}:${lib.makeBinPath [ builder ]}:"$PATH"
+          SHELL=${lib.escapeShellArg shell}
+          BASH=${lib.escapeShellArg shell}
+          set +e
+          [ -n "$PS1" -a -z "$NIX_SHELL_PRESERVE_PROMPT" ] && PS1='\n\[\033[1;32m\][nix-shell:\w]\$\[\033[0m\] '
+          if [ "$(type -t runHook)" = function ]; then
+            runHook shellHook
+          fi
+          unset NIX_ENFORCE_PURITY
+          shopt -u nullglob
+          shopt -s execfail
+          ${optionalString (command != null || run != null) ''
+            ${optionalString (command != null) command}
+            ${optionalString (run != null) run}
+            exit
+          ''}
+        '';
+
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libstore/globals.hh#L464-L465
+        sandboxBuildDir = "/build";
+
+        setupPackage = runCommandNoCC "setup" {
+          # https://github.com/NixOS/nix/blob/ffe155abd36366a870482625543f9bf924a58281/src/libstore/build/local-derivation-goal.cc#L906-L910
+          passwd = ''
+            root:x:0:0:Root user:${sandboxBuildDir}:/noshell
+            nixbld:x:${toString uid}:${toString gid}:Build user:${sandboxBuildDir}:/noshell
+            nobody:x:65534:65534:Nobody:/:/noshell
+          '';
+          group = ''
+            root:x:0:
+            nixbld:!:${toString gid}:
+            nogroup:x:65534:
+          '';
+          passAsFile = [ "passwd" "group" ];
+        } ''
+          # Directories that are later made writable in fakeRootCommands
+          mkdir -p "$out"{/tmp,${storeDir},${sandboxBuildDir}}
+
+          # Directories for internet access and POSIX compat
+          mkdir -p "$out"{/etc/ssl/certs,/bin,/usr/bin}
+          # Allows Nix binaries to find CA certificates for internet access
+          ln -s ${cacert}/etc/ssl/certs/ca-bundle.crt "$out"/etc/ssl/certs/ca-certificates.crt
+          # POSIX compat, for shell scripts and co.
+          ln -s ${bashInteractive}/bin/sh "$out"/bin/sh
+          ln -s ${coreutils}/bin/env "$out"/usr/bin/env
+          # Allows programs to look up the build users home directory
+          cp "$passwdPath" "$out"/etc/passwd
+          cp "$groupPath" "$out"/etc/group
+        '';
+
+        # This function closely mirrors what this Nix code does:
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/primops.cc#L1102
+        # https://github.com/NixOS/nix/blob/2.8.0/src/libexpr/eval.cc#L1981-L2036
+        stringValue = value:
+          # We can't just use `toString` on all derivation attributes because that
+          # would not put path literals in the closure. So we explicitly copy
+          # those into the store here
+          if builtins.typeOf value == "path" then "${value}"
+          else if builtins.typeOf value == "list" then toString (map stringValue value)
+          else toString value;
```

Comment:
This smells like it might be useful outside of `pkgs/build-support/docker/default.nix`. @infinisil, @roberth , any thoughts?

Reply:
It does smell.

The name should be as awful as the concept of automatic string coercions. Nah, but more seriously maybe reflect the "coerceMore = true" param a bit:

`moreToString`?

I can't get excited about this function, but I guess it needs to be done. I would be surprised if there isn't already another implementation somewhere. Probably a non-recursive one because the use case didn't need it. Probably many if you include all the degenerate ones.

Reply:
Actually recursion is kind of needed, because you can also do `var = [ "a" [ "b" [ "c" ] ] ];` in derivations.

I think for now this can stay localized here, I don't remember any need for this function anywhere else. Ideally this whole PR would be done in Nix itself, because so much functionality of Nix is straight-up copied here. But I think it's nice as a starter.

Reply:
>  Ideally this whole PR would be done in Nix itself, because so much functionality of Nix is straight-up copied here.

Anything you do in Nix itself is not pinnable and needs extra care for compatibility. I'd take an expression library over complex builtins any day. Take the fetching builtins for instance, they're almost impossible to fix without breaking someone's code irreversibly, as their SHAs will be wrong.

Reply:
> stay localized here

sgtm